lib/charms/smtp_integrator/v0/smtp.py

```diff
@@ -0,0 +1,344 @@
+    """SMTP events.
+
+    This class defines the events that a SMTP requirer can emit.
+
+    Attributes:
+        smtp_data_available: the SmtpDataAvailableEvent.
+    """
+
+    smtp_data_available = ops.EventSource(SmtpDataAvailableEvent)
+
+
+class SmtpRequires(ops.Object):
+    """Requirer side of the SMTP relation.
+
+    Attributes:
+        on: events the provider can emit.
+    """
+
+    on = SmtpRequiresEvents()
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str = DEFAULT_RELATION_NAME) -> None:
+        """Construct.
+
+        Args:
+            charm: the provider charm.
+            relation_name: the relation name.
+        """
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    def get_relation_data(self) -> Optional[SmtpRelationData]:
+        """Retrieve the relation data.
+
+        Returns:
+            SmtpRelationData: the relation data.
+        """
+        relation = self.model.get_relation(self.relation_name)
+        return self._get_relation_data_from_relation(relation) if relation else None
+
+    def _get_relation_data_from_relation(self, relation: ops.Relation) -> SmtpRelationData:
+        """Retrieve the relation data.
+
+        Args:
+            relation: the relation to retrieve the data from.
+
+        Returns:
+            SmtpRelationData: the relation data.
+        """
+        assert relation.app
+        relation_data = relation.data[relation.app]
+        return SmtpRelationData(
+            host=relation_data.get("host"),
+            port=relation_data.get("port"),
+            user=relation_data.get("user"),
+            password=relation_data.get("password"),
+            password_id=relation_data.get("password_id"),
+            auth_type=relation_data.get("auth_type"),
+            transport_security=relation_data.get("transport_security"),
+            domain=relation_data.get("domain"),
+        )
+
+    def _is_relation_data_valid(self, relation: ops.Relation) -> bool:
+        """Validate the relation data.
+
+        Args:
+            relation: the relation to validate.
+
+        Returns:
+            true: if the relation data is valid.
+        """
+        try:
+            _ = self._get_relation_data_from_relation(relation)
+            return True
+        except ValidationError:
+            return False
+
+    def _on_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        """Event emitted when the relation has changed.
+
+        Args:
+            event: event triggering this handler.
+        """
+        assert event.relation.app
+        relation_data = event.relation.data[event.relation.app]
+        if relation_data:
+            if relation_data["auth_type"] == AuthType.NONE.value:
+                logger.warning('Insecure setting: auth_type has a value "none"')
```

Comment:
Our thinking was that the provider charm operator should be aware of how to configure the charm, however, the requirer relies blindly on the provider config, hence the warning

Reply:
Yeah, got it. I still think it would make sense to add a similar warning on a provider side tho. Operator should be aware of how to configure things, but operators also make mistakes ;)

Reply:
I've just merged https://github.com/canonical/smtp-integrator-operator/pull/33. I think we can close this

Reply:
Yup, looks good.